### ISSUE TITLE
Sawtooth artifacts viewing an image with super high zoom levels

### ISSUE THREAD

**Issue by pwinston** [vispy/vispy#1852]
When I run `examples/basics/scene/image.py` and zoom *way* into the image I see a "sawtooth" artifacts at the boundary between two image pixels as shown here:

<img width="795" alt="vispy-sawtooth" src="https://user-images.githubusercontent.com/4163446/80154531-c332f100-858d-11ea-816d-5ea25ce84cfc.png">

I created my own test program so I could print out the "zoom factor". A zoom factor of 1.0 means a single image pixel is the size of the window. Below is zoom 432 when the artifacts are just barely visible as nearly single pixel noise on vertical edge:

<img width="600" alt="vispy-zoom-432" src="https://user-images.githubusercontent.com/4163446/80154935-8ddad300-858e-11ea-8842-67a785a0a457.png">

To get the really big sawtooths here I'm using zoom 132000:

<img width="598" alt="vispy-zoom-123000" src="https://user-images.githubusercontent.com/4163446/80154991-aea32880-858e-11ea-9333-d59913897465.png">

A very reasonable question is who is ever going to zoom in this far! We noticed this in https://github.com/napari/napari. Although we can probably limit the zoom level, we do support having multiple layers of images at very different resolutions. So it's not crazy to be have an extremely coarse image and need to zoom inside a single image pixel.

Above images were on a 2019 MacBook Pro with AMD 5300M 4GB gpu.

Using a Windows laptop with NVIDIA 980M gpu the problem is *much* less pronounced. I never see full sawtooths but around zoom = 70,000 I do see some weirdness.

Setting the vispy image method to 'impostor' instead of 'subdivide' fixes the problem. But as the docs say this could be slower, so we're not eager to switch to this method.

**Comment by djhoese** [comment]
Interesting. I ran your zoom_test on my Ubuntu machine and didn't see the issue at all. Am I doing something wrong?

![image](https://user-images.githubusercontent.com/1828519/80222562-aa1d5500-860c-11ea-9575-cec01318466c.png)

```
In [2]: print(vispy.sys_info())                                                                                                                                                                                                                                     
Platform: Linux-5.3.0-7648-generic-x86_64-with-debian-buster-sid
Python:   3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 21:52:21)  [GCC 7.3.0]
NumPy:    1.17.2
Backend:  PyQt5
pyqt4:    None
pyqt5:    ('PyQt5', '5.12.3', '5.12.5')
pyside:   None
pyside2:  None
pyglet:   None
glfw:     None
sdl2:     None
wx:       None
egl:      EGL 1.5 NVIDIA: OpenGL_ES OpenGL
osmesa:   None
_test:    None

GL version:  '4.6.0 NVIDIA 440.64'
MAX_TEXTURE_SIZE: 32768
```

**Comment by djhoese** [comment]
Ah you have impostor set. Retrying...

**Comment by djhoese** [comment]
That's more like it:

![image](https://user-images.githubusercontent.com/1828519/80223319-bf46b380-860d-11ea-941f-c040a22a0ee1.png)

**Comment by djhoese** [comment]
If I had to make a wild guess I would say we're dealing with floating point precision on the GPU. Perhaps in the vertex/texture coordinate calculations in Python (see `ImageVisual._build_vertex_data`) but even then the buffers on the lower levels of VisPy need to be 32-bit floats.

**Comment by pwinston** [comment]
My mistake on the test program, I swapped it back to 'subdivide'. Your image is exactly how my NVIDIA laptop looks. As my images show it's much worse on AMD. Although even on AMD we are talking about pretty extreme zoom levels.

I agree it has to be precision related. But I suspect it's not "inevitable", like if it were just a known thing that all programs had this artifact at extreme zooms I suspect it would be documented. It'd be a well known thing "artifacts appear if you zoom too much, so don't zoom too much". 

For napari we aren't too concerned about this issue. I think long term we want limit zoom anyway, for multiple reason. But for now we're just going to do nothing and see if it keeps coming up. We wanted to document it though. Maybe someone will see this and go "oh that, yeah you just need to...".

**Comment by tlambert03** [comment]
@djhoese, I wonder if this is another reason to get on that PR I was [thinking about ](https://github.com/vispy/vispy/pull/1842#issuecomment-601235221) to open up access to other texture bit depths? 😅

@pwinston, as a quick test, i wonder if it might help confirm any of your theories if you changed the hard-coded 8-bit texture bit depth to something like `GL_FLOAT` in the 5th argument to `glTexImage3D` in [this line](https://github.com/vispy/vispy/blob/cf244a266fe71a5b306ef518f0f8a05040502939/vispy/gloo/glir.py#L1623-L1624) (for 3D) or [this one](https://github.com/vispy/vispy/blob/cf244a266fe71a5b306ef518f0f8a05040502939/vispy/gloo/glir.py#L1527-L1528) (for 2D)?

**Comment by djhoese** [comment]
@tlambert03 I don't think this is texture bit depth related, but rather vertex and/or texture *coordinate* related. Changing the data type (bit depth) of the texture shouldn't effect this. Changing that should only effect the color/quality of the image.

I'm going to see if I can change the ImageVisual to use larger vertex coordinates instead of sub-1 and see what happens.

**Comment by tlambert03** [comment]
ahhh.  my bad :)

**Comment by djhoese** [comment]
Ok I still need to play around with this more, but I found a bug in the ImageVisual. Sadly, this has no effect on the problems in this issue. The bug is that the `_need_vertex_update` flag is never set to `False` in `_build_vertex_data` so the coordinates are computed every time. Adding a `self._need_vertex_update = False` to the end of that method fixes that.

I looked at the differences between impostor and subdivide and I'm surprised by how different the two code segments are. The subdivide method does a traditional vertex coordinate transformation in the vertex shader. The impostor method does it in the fragment shader along with some more complex math. This makes me wonder if the camera transformation is the problem. But that would mean it is bad in the forward path but fine in the inverse transformation (what impostor uses in the fragment shader).

**Comment by djhoese** [comment]
So I changed the bottom of the test script to:

```
image = scene.visuals.Image(checkerboard(), interpolation='nearest',
                            parent=view.scene, method='subdivide')  #, method='impostor')

from vispy.visuals.transforms import STTransform
image.transform = STTransform(scale=(360000., 360000.), translate=(-359700., -359700.))


# view.camera = TestPanZoomCamera(aspect=1)
# flip y-axis to have correct alignment
# view.camera.flip = (0, 1, 0)
# view.camera.set_range()
# view.camera.zoom(0.1, (1, 1))
```

And I don't see the issue. Does this mean anything? No idea. I've been looking at this too long.

**Comment by pwinston** [comment]
I looked at this just to learn a bit of vispy, not because we really care about this too much.

I think the problem is float precision in the scene matrix, for me it looks like this:

    [5099842.5, 0.0, 0.0, 0.0],
    [0.0, 5099842.5, 0.0, 0.0],
    [0.0, 0.0, 9.999999974752427e-07, 0.0],
    [-5089572.0, -5089587.0, 0.0, 1.0]]

In vispy/numpy it's float64 but I believe in the card it will be 32-bit floats. You can do 64-bit stuff with GLSL extensions but I'd strongly doubt we'd want to switch to 64-bit for this extremely fringe case.

The big numbers above come up just because you are mapping [-tiny..tiny] (100,000 times smaller than a pixel) to to the window [0...600]. Big numbers are not a problem for floats, but above we scale by 5099842.5 and then slide it over by -5089572.0. You only have ~7 digits of precision so that won't be stable. The verts and texcoords get multiplied by this matrix. This is hand wavy, but I think the above matrix is inherently bad news.

I don't quite understand the impostor mode, but it works. I think it must avoid the massive dynamic range issue, because it basically refers to just a portion of the full texture. Which in a sense is like you'd do with any sort of multi-scale viewer, you essentially have tiles, so when rendering super close up you only refer to a view small tiles and not the full image.

At any rate I think this is all massively out of scope for vispy. I think we already have multi-scale support in napari. The counter-intuitive thing here is you need multi-scale even though your original image is not big, but you need it because you are navigating/zooming in a multi-scale way.

And if we did just limit the zoom so you can pixel peep, see individual pixels, but not see 1/100,000th of a pixel, that probably would be sufficient for now. Kind of a wild goose chance but was interesting for me. Stepping through in the debugger was very helpful.

**Comment by pwinston** [comment]
Thinking about it, vispy could have a new Image mode 'auto' which does 'subdivide' under normal conditions, since that's faster, but swaps to 'impostor' mode if the scene transform is in dangerous territory, to avoid artifacts. That would in some ways be the best of both worlds.

 Very not sure that's worth doing, but just to mention it.

**Comment by djhoese** [comment]
Thanks for the analysis/description. Sounds very reasonable.

> I think we already have multi-scale support in napari.

Do you? What does this mean exactly? Is this something that could be ported to vispy?

**Comment by pwinston** [comment]
See the "image pyramids" heading on this page for an animation:
https://napari.org/tutorials/fundamentals/image

Basically it means you have an image that's much larger than the maximum texture size, say 50,000 x 50,000 pixels. So you build a hierarchy of "tiles" like with Google Maps, where at the finest resolution the tiles are 1:1 the full resolution, but each coarser set of tiles are downsampled more and more, and you have less and less tiles at each level. The net result is you can draw the image at any zoom level or pan location just by drawing a handful of tiles. And this scales pretty much infinitely as Google Maps shows. I think Google Maps has just 20 levels to cover the whole earth.

I'm really brand new to napari so I can't begin to say if this is something that should ever be moved to vispy or not. I can ask around. I'm hopefully going to work on napari full time for a few months starting in May so I'll hopefully learn more.
